File: example implementations/java/convertor/src/io/ColourMap.java
package io;

import utils.EL;
import voxel.Colour;
import static utils.NumberUtilities.*;

import java.util.HashMap;

public class ColourMap {
    EL<Colour> order = new EL<>();
    HashMap<Colour, Integer> counts = new HashMap<>();
    Colour current;
    int colourPointer = 0;
    int countPointer = 0;
    int countTotal = 0;

    void add(Colour colour, int count) {
        order.add(colour);
        counts.put(colour, count);
    }

    EL<Colour> getUnused() {
        EL<Colour> res = new EL<>();
        for (Colour colour : counts.keySet())
            if (counts.get(colour) == 0)
                res.add(colour);
        return res;
    }

    /**
     * @return Next colour if possible, null otherwise.
     */
    Colour getNext() {
        if (countPointer == countTotal) {
            countPointer = 0;
            getNew();
        }
        countPointer++;
        return current;
    }

    /**
     * Try to advance the current colour, set it to null if not possible.
     */
    private void getNew() {
        if (colourPointer < order.size()) {
            current = order.get(colourPointer);
            countTotal = counts.get(current);
            colourPointer++;
        }
        else
            current = null;
    }

    public byte[] toBytes()
    {
        byte[] res = new byte[order.size() * 7];
        for (int cc = 0; cc < order.size(); cc++)
        {
            Colour colour = order.get(cc);
            System.arraycopy(intsToBytes(Colour.argbToRgba(colour.getRGB())), 0, res, cc * 7, 4);
            System.arraycopy(intsToBytes(3, true, counts.get(colour)), 0, res, cc * 7 + 4, 3);
        }
        return res;
    }

}

File: example implementations/csharp/cvox-convertor/io/ColourMap.cs
using static cvox_convertor.utils.Extensions;
using static cvox_convertor.utils.NumberUtilities;
using System.Drawing;

namespace cvox_convertor.io
{
    public class ColourMap
    {
        List<Color> Order = new();
        Dictionary<Color, int> Counts = new();
        Color current;
        int colourPointer = 0;
        int countPointer = 0;
        int countTotal = 0;
        bool EndReached = false;

        public void Add(Color colour, int count)
        {
            Order.Add(colour);
            Counts.Add(colour, count);
        }

        public List<Color> GetUnused()
        {
            List<Color> res = new();
            foreach (Color colour in Counts.Keys)
                if (Counts[colour] == 0)
                    res.Add(colour);
            return res;
        }

        /**
         * @return Next colour if possible, null otherwise.
         */
        public Color? GetNext()
        {
            if (countPointer == countTotal)
            {
                countPointer = 0;
                GetNew();
            }
            countPointer++;
            return EndReached ? null : current;
        }

        public byte[] ToBytes()
        {
            byte[] res = new byte[Order.Count * 7];
            for (int cc = 0; cc < Order.Count; cc++)
            {
                Color colour = Order[cc];
                Array.Copy(IntsToBytes(colour.ToRgba()), 0, res, cc * 7, 4);
                Array.Copy(IntsToBytes(3, true, Counts[colour]), 0, res, cc * 7 + 4, 3);
            }
            return res;
        }

        /**
         * Try to advance the current colour.
         */
        private void GetNew()
        {
            if (colourPointer < Order.Count)
            {
                current = Order[colourPointer];
                countTotal = Counts[current];
                colourPointer++;
            }
            else
                EndReached = true;
        }

    }
}
